### ISSUE TITLE
Switches and Checkboxes on iOS have accessibility issues

### ISSUE THREAD

**Issue by jkurtw** [flutter/flutter#49259]
**Internal:** b/148045557

Switches and Checkboxes on iOS have accessibility issues.

In the following app / screenshot:
![checkboxes](https://user-images.githubusercontent.com/49886147/72854825-6eb86680-3c6a-11ea-8620-fba31a7da70c.png)

**Seen result:**
Tapping on 'Switch' just says 'Switch'.
Tapping on 'Checkbox' just says 'Checkbox'.

There's no indication that the field is on or off, or that it can be toggled with a double tap.

If the value is on, then this reads:

'Selected, Switch'
'Selected, Checkbox'

**Expected result:**

Based on how this works with normal iOS switches, I'd expect:

'Switch, off, double tap to toggle setting'
'Checkbox, off, double tap to toggle setting'

If they're on, I'd expect:

'Switch, on, double tap to toggle setting'
'Checkbox, on, double tap to toggle setting'

**Full repro code:**

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  State createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  bool _on = false;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Accessibility test',
      theme: ThemeData(
        colorScheme: ColorScheme.light(),
        primarySwatch: Colors.blue,
      ),
      darkTheme: ThemeData(
        colorScheme: ColorScheme.dark(),
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(title: Text('Accessibility test')),
        body: Column(
          children: [
            SwitchListTile(
              title: Text('Switch'),
              value: _on,
              onChanged: (value) => setState(() => _on = value),
            ),
            CheckboxListTile(
              title: Text('Checkbox'),
              value: _on,
              onChanged: (value) => setState(() => _on = value),
            )
          ],
        ),
      ),
    );
  }
}
```

**Comment by goderbauer** [comment]
Checkboxes should probably say "checked" or "unchecked" instead of on/off.

We could probably put the on/off/checked/unchecked string into the value field of the SemanticsBridge and have the iOS bridge read that out (instead of adding the selected AX trait).

This may just be a dupe of https://github.com/flutter/flutter/issues/16395. /cc @jonahwilliams 

I don't think we can provide the "double tap to toggle setting" hint in a meaningful way, though. Depending on what a11y service somebody uses, the activation action may be different.

**Comment by jonahwilliams** [comment]
Potentially solved with workaround listed in second answer of https://stackoverflow.com/questions/52230673/how-can-i-make-an-arbitrary-uiaccessibilityelement-behave-like-a-uiswitch-for-vo/52364529

**Comment by jonahwilliams** [comment]
Also see: https://github.com/jldailey13/24-Accessibility-Article-Examples/blob/master/24A11y%20Article/Name%20Examples/AccessibleSwitchView.swift

**Comment by goderbauer** [comment]
Ah, right! We can just steal the trait. I think we should do that and see if that gets us the behavior we want. That could work well for the Switch.

We'd still have to do something custom for checkboxes (If we actually want them to announce something when they're unchecked).

**Comment by goderbauer** [comment]
Also according to https://material.io/design/platform-guidance/cross-platform-adaptation.html#cross-platform-guidelines Checkboxes shouldn't really be used on iOS (which kinda makes sense since the system doesn't have an a11y trait for them...)

**Comment by gaaclarke** [comment]
The reason it says "selected" right now is because of this code:

https://github.com/flutter/engine/blob/4a55b4fd2500bf96de93c4feb4172d88b93d52c3/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L430:L430

It is setting the `UIAccessibilityTraitSelected` trait which according to Apple's document as:
```
 Used when the element is selected.
 For example, a selected row in a table or a selected button within a segmented control.
```

That doesn't quite match up with the meaning of a switch being on or off.

**Comment by goderbauer** [comment]
Yeah, iOS doesn't have a public trait for a switch. The idea is to instantiate an actual UIKit Switch and steal the a11y trait from it as suggested in one of the links posted by Jonah.

We should do the same for cehckboxes and switches and then document on checkboxes that they shouldn't be used on iOS with a link to the Material Spec linked above.

**Comment by gaaclarke** [comment]
The proper place to specify if a switch is off or on would be here:

https://github.com/flutter/engine/blob/4a55b4fd2500bf96de93c4feb4172d88b93d52c3/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L265

The problem is that the text needs to be localized.  Using a standard UISwitch you get the following:
```
(lldb) p (int)[ui accessibilityTraits]
(int) $1 = 1
(lldb) po (NSString*)[ui accessibilityValue]
1
(lldb) po (NSString*)[ui accessibilityHint]
 nil
```

Returning "1" from accessibilityValue causes the VoiceOver to say "one" not "on".

**Comment by gaaclarke** [comment]
@goderbauer Thank, I'm playing around with that.  I haven't got it to work yet.

**Comment by gaaclarke** [comment]
It also looks that if you use UISwitch's accessibilityTraits, VoiceOver says: "<label>, <value>, button" . That may be a way to avoid trying be explicit about what interaction you need to take to change it, just label it a button.

**Comment by gaaclarke** [comment]
I attempted to delegate to a real UISwitch for `accessibilityTraits` `accessibilityValue` `accessibilityHint` and `accessibilityLabel`, VoiceOver would just say 'zero, button' 'one, button'.

It did not say anything about taps, and it didn't say 'on' or 'off'.

```diff
diff --git a/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm b/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
index 86cec6704..790548337 100644
--- a/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
+++ b/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
@@ -93,8 +93,13 @@ flutter::SemanticsAction GetSemanticsActionForScrollDirection(
 
 @end
 
+@interface SemanticsObject ()
+@property (nonatomic, readonly) UISwitch* uiSwitch;
+@end
+
 @implementation SemanticsObject {
   fml::scoped_nsobject<SemanticsObjectContainer> _container;
+  UISwitch *_uiSwitch;
 }
 
 #pragma mark - Override base class designated initializers
@@ -135,11 +140,19 @@ flutter::SemanticsAction GetSemanticsActionForScrollDirection(
   _parent = nil;
   _container.get().semanticsObject = nil;
   [_platformViewSemanticsContainer release];
+  [_uiSwitch release];
   [super dealloc];
 }
 
 #pragma mark - Semantic object methods
 
+- (UISwitch*)uiSwitch {
+  if (!_uiSwitch) {
+    _uiSwitch = [[UISwitch alloc] init];
+  }
+  return _uiSwitch;
+}
+
 - (BOOL)isAccessibilityBridgeAlive {
   return [self bridge].get() != nil;
 }
@@ -248,6 +261,10 @@ flutter::SemanticsAction GetSemanticsActionForScrollDirection(
   if (![self isAccessibilityBridgeAlive])
     return nil;
 
+  if ([self node].HasFlag(flutter::SemanticsFlags::kHasToggledState)) {
+    self.uiSwitch.on = [self node].HasFlag(flutter::SemanticsFlags::kIsToggled);
+    return self.uiSwitch.accessibilityLabel;
+  }
   if ([self node].label.empty())
     return nil;
   return @([self node].label.data());
@@ -257,6 +274,10 @@ flutter::SemanticsAction GetSemanticsActionForScrollDirection(
   if (![self isAccessibilityBridgeAlive])
     return nil;
 
+  if ([self node].HasFlag(flutter::SemanticsFlags::kHasToggledState)) {
+    self.uiSwitch.on = [self node].HasFlag(flutter::SemanticsFlags::kIsToggled);
+    return self.uiSwitch.accessibilityHint;
+  }
   if ([self node].hint.empty())
     return nil;
   return @([self node].hint.data());
@@ -265,6 +286,11 @@ flutter::SemanticsAction GetSemanticsActionForScrollDirection(
 - (NSString*)accessibilityValue {
   if (![self isAccessibilityBridgeAlive])
     return nil;
+
+  if ([self node].HasFlag(flutter::SemanticsFlags::kHasToggledState)) {
+    self.uiSwitch.on = [self node].HasFlag(flutter::SemanticsFlags::kIsToggled);
+    return self.uiSwitch.accessibilityValue;
+  }    
   if ([self node].value.empty())
     return nil;
   return @([self node].value.data());
@@ -424,7 +450,10 @@ flutter::SemanticsAction GetSemanticsActionForScrollDirection(
     traits |= UIAccessibilityTraitAdjustable;
   }
   // TODO(jonahwilliams): switches should have a value of "on" or "off"
-  if ([self node].HasFlag(flutter::SemanticsFlags::kIsSelected) ||
+  if ([self node].HasFlag(flutter::SemanticsFlags::kHasToggledState)) {
+    self.uiSwitch.on = [self node].HasFlag(flutter::SemanticsFlags::kIsToggled);
+    return self.uiSwitch.accessibilityTraits;
+  } else if ([self node].HasFlag(flutter::SemanticsFlags::kIsSelected) ||
       [self node].HasFlag(flutter::SemanticsFlags::kIsToggled) ||
       [self node].HasFlag(flutter::SemanticsFlags::kIsChecked)) {
     traits |= UIAccessibilityTraitSelected;
```

**Comment by jonahwilliams** [comment]
This doesn't include forcing a 0 or 1 label like the linked code

**Comment by gaaclarke** [comment]
> This doesn't include forcing a 0 or 1 label like the linked code

Where do you see someone doing that?  That patch is just returning the label that UISwitch would return currently.

**Comment by jonahwilliams** [comment]
Sorry, the value not the label, see my second link

**Comment by gaaclarke** [comment]
The line `return self.uiSwitch.accessibilityValue;` is returning "1" or "0" that's why it is saying that.  It doesn't get mapped to "on" and "off" for me.

**Comment by gaaclarke** [comment]
I used objective-c method forwarding to send all messages to a real UISwitch except for `accessibilityFrame` and VoiceOver still read "0, Button".  Whatever magic UISwitch does to get VoiceOver to say ("on" | "off" ) isn't accessible from the UISwitch.

```diff
diff --git a/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm b/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
index 86cec6704..8b172d10d 100644
--- a/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
+++ b/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
@@ -14,6 +14,40 @@
 #include "flutter/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h"
 #include "flutter/shell/platform/darwin/ios/platform_view_ios.h"
 
+@interface Foobar : UISwitch
+@end
+
+@implementation Foobar {
+  SemanticsObject* _semanticsObject;
+}
+
+- (instancetype)initWithSemanticsObject:(SemanticsObject*)semanticsObject {
+  self = [super init];
+  if (self) {
+    _semanticsObject = [semanticsObject retain];
+  }
+  return self;
+}
+
+- (NSMethodSignature *)methodSignatureForSelector:(SEL)sel {
+  NSMethodSignature *result = [super methodSignatureForSelector:sel];
+  if (!result) {
+    result = [_semanticsObject methodSignatureForSelector:sel];
+  }
+  return result;
+}
+
+- (void)forwardInvocation:(NSInvocation *)anInvocation {
+  [anInvocation setTarget:_semanticsObject];
+  [anInvocation invoke];
+}
+
+- (CGRect)accessibilityFrame {
+  return [_semanticsObject accessibilityFrame];
+}
+
+@end
+
 namespace {
 
 constexpr int32_t kRootNodeId = 0;
@@ -765,6 +799,8 @@ SemanticsObject* AccessibilityBridge::GetOrCreateObject(int32_t uid,
         !node.HasFlag(flutter::SemanticsFlags::kIsReadOnly)) {
       // Text fields are backed by objects that implement UITextInput.
       object = [[[TextInputSemanticsObject alloc] initWithBridge:GetWeakPtr() uid:uid] autorelease];
+    } else if (node.HasFlag(flutter::SemanticsFlags::kHasToggledState) ){
+      object = (SemanticsObject*)[[Foobar alloc] initWithSemanticsObject:[[[FlutterSemanticsObject alloc] initWithBridge:GetWeakPtr() uid:uid] autorelease]];
     } else {
       object = [[[FlutterSemanticsObject alloc] initWithBridge:GetWeakPtr() uid:uid] autorelease];
     }
```

**Comment by gaaclarke** [comment]
I created an empty iOS project and looked at what a standard iOS app will do.  For a switch it says: "_label_, _("0" | "1")_, button".  I will duplicate that behavior.

**Comment by gaaclarke** [comment]
fixed in https://github.com/flutter/engine/pull/16211
